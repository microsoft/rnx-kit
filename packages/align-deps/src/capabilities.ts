import type { Capability } from "@rnx-kit/config";
import { warn } from "@rnx-kit/console";
import { keysOf } from "@rnx-kit/tools-language/properties";
import type { PackageManifest } from "@rnx-kit/tools-node/package";
import type { MetaPackage, Package, Preset, Profile } from "./types";

type ResolvedDependencies = {
  dependencies: Record<string, Package[]>;
  unresolvedCapabilities: Record<string, string[]>;
};

/**
 * Returns the list of capabilities used in the specified package manifest.
 * @param packageManifest The package manifest to scan for dependencies
 * @param preset The preset to use to resolve capabilities
 * @returns A list of capabilities used in the specified package manifest
 */
export function capabilitiesFor(
  {
    dependencies = {},
    devDependencies = {},
    peerDependencies = {},
  }: PackageManifest,
  preset: Preset
): Capability[] {
  const dependenciesSet = new Set<string>(Object.keys(dependencies));
  Object.keys(peerDependencies).forEach((dep) => dependenciesSet.add(dep));
  Object.keys(devDependencies).forEach((dep) => dependenciesSet.add(dep));

  if (dependenciesSet.size === 0) {
    return [];
  }

  const foundCapabilities = new Set<Capability>();
  for (const profile of Object.values(preset)) {
    for (const capability of keysOf(profile)) {
      const { name } = profile[capability];
      if (dependenciesSet.has(name)) {
        foundCapabilities.add(capability);
      }
    }
  }

  return Array.from(foundCapabilities).sort();
}

export function isMetaPackage(pkg: MetaPackage | Package): pkg is MetaPackage {
  return pkg.name === "#meta" && Array.isArray(pkg.capabilities);
}

function resolveCapability(
  capability: Capability,
  namedProfile: [string, Profile],
  dependencies: Record<string, Package[]>,
  unresolvedCapabilities: Record<string, string[]>,
  resolved = new Set<string>()
): void {
  if (resolved.has(capability)) {
    return;
  }

  // Make sure we don't end in a loop
  resolved.add(capability);

  const [profileName, profile] = namedProfile;
  const pkg = profile[capability];
  if (!pkg) {
    const profiles = unresolvedCapabilities[capability];
    if (!profiles) {
      unresolvedCapabilities[capability] = [profileName];
    } else {
      profiles.push(profileName);
    }
    return;
  }

  pkg.capabilities?.forEach((capability) =>
    resolveCapability(
      capability,
      namedProfile,
      dependencies,
      unresolvedCapabilities,
      resolved
    )
  );

  if (!isMetaPackage(pkg)) {
    const { name, version } = pkg;
    if (name in dependencies) {
      const versions = dependencies[name];
      if (!versions.find((current) => current.version === version)) {
        versions.push(pkg);
      }
    } else {
      dependencies[name] = [pkg];
    }
  }
}

/**
 * Resolves specified capabilities to real dependencies.
 * @param capabilities The list of capabilities to resolve
 * @param preset The preset to use to resolve capabilities
 * @returns A tuple of resolved dependencies and unresolved capabilities
 */
export function resolveCapabilitiesUnchecked(
  capabilities: Capability[],
  preset: Preset
): ResolvedDependencies {
  const profiles = Object.entries(preset);
  const dependencies: Record<string, Package[]> = {};
  const unresolvedCapabilities: Record<string, string[]> = {};

  for (const capability of capabilities) {
    profiles.forEach((profile) => {
      resolveCapability(
        capability,
        profile,
        dependencies,
        unresolvedCapabilities
      );
    });
  }

  return { dependencies, unresolvedCapabilities };
}

/**
 * Resolves specified capabilities to real dependencies.
 *
 * Same as {@link resolveCapabilitiesUnchecked}, but warns about any unresolved
 * capabilities.
 *
 * @param manifestPath The path to the package manifest
 * @param capabilities The list of capabilities to resolve
 * @param preset The preset to use to resolve capabilities
 * @returns Resolved dependencies
 */
export function resolveCapabilities(
  manifestPath: string,
  capabilities: Capability[],
  preset: Preset
): Record<string, Package[]> {
  const { dependencies, unresolvedCapabilities } = resolveCapabilitiesUnchecked(
    capabilities,
    preset
  );

  const unresolved = Object.entries(unresolvedCapabilities);
  if (unresolved.length > 0) {
    const message = unresolved.reduce(
      (lines, [capability, profiles]) =>
        (lines += `\n\t${capability} (missing in ${profiles.join(", ")})`),
      `${manifestPath}: The following capabilities could not be resolved for one or more profiles:`
    );

    warn(message);
  }

  return dependencies;
}
