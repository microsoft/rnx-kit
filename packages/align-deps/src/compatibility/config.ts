import type { KitConfig } from "@rnx-kit/config";
import { warn } from "@rnx-kit/console";
import { defaultConfig } from "../config";
import { dropPatchFromVersion, modifyManifest } from "../helpers";
import type { AlignDepsConfig, LegacyCheckConfig, Options } from "../types";

const legacyKeys = [
  "capabilities",
  "customProfiles",
  "reactNativeDevVersion",
  "reactNativeVersion",
] as const;

function findLegacyConfigKeys(config: KitConfig): (keyof KitConfig)[] {
  return legacyKeys.filter((key) => key in config);
}

/**
 * Transforms the old config schema into the new one.
 *
 * Note that this config is presented to the user and should therefore be
 * "pretty".
 *
 * @param oldConfig Config in old schema
 * @returns Config in new schema
 */
export function transformConfig({
  capabilities,
  customProfiles,
  kitType,
  manifest,
  reactNativeDevVersion,
  reactNativeVersion,
}: LegacyCheckConfig): AlignDepsConfig {
  const devVersion = dropPatchFromVersion(
    reactNativeDevVersion || reactNativeVersion
  );
  const prodVersion = dropPatchFromVersion(reactNativeVersion);
  return {
    kitType,
    alignDeps: {
      presets: customProfiles
        ? [...defaultConfig.presets, customProfiles]
        : defaultConfig.presets,
      requirements:
        kitType === "app"
          ? [`react-native@${prodVersion}`]
          : {
              development: [`react-native@${devVersion}`],
              production: [`react-native@${prodVersion}`],
            },
      capabilities,
    },
    manifest,
  };
}

/**
 * Migrates the old config schema into the new one, if necessary.
 *
 * This will function will allow users to let `align-deps` update their config.
 * Otherwise, it will tell them how to update their config manually. It will
 * also warn about old config keys that are no longer used.
 *
 * @param config Configuration in the package manifest
 * @param manifestPath Path to the package manifest to check
 * @param options Command line options
 * @returns The config in the new schema
 */
export function migrateConfig(
  config: AlignDepsConfig | LegacyCheckConfig,
  manifestPath: string,
  { migrateConfig }: Options
): AlignDepsConfig {
  if ("alignDeps" in config) {
    const oldKeys = findLegacyConfigKeys(config);
    if (oldKeys.length > 0) {
      const unsupportedKeys = oldKeys
        .map((key) => `'rnx-kit.${key}'`)
        .join(", ");
      warn(
        `${manifestPath}: The following keys are no longer supported: ${unsupportedKeys}`
      );
    }
    return config;
  }

  const newConfig = transformConfig(config);
  const { manifest, ...configOnly } = newConfig;

  if (migrateConfig) {
    const kitConfig = manifest["rnx-kit"];
    if (kitConfig) {
      for (const key of legacyKeys) {
        delete kitConfig[key];
      }
    }

    manifest["rnx-kit"] = {
      ...kitConfig,
      alignDeps: configOnly.alignDeps,
    };

    modifyManifest(manifestPath, manifest);
  } else {
    warn(`${manifestPath}: The config schema has changed. Please update your config to the following:

${JSON.stringify(configOnly, null, 2)}

Or run this command again with '--migrate-config' to update your config automatically.

Support for the old schema will be removed in a future release.`);
  }

  return newConfig;
}
