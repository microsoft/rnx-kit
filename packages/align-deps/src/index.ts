export { capabilitiesFor } from "./capabilities";
export { checkPackageManifest } from "./check";
export { cli } from "./cli";
export { updatePackageManifest } from "./manifest";

export type {
  Args,
  CapabilitiesOptions,
  CheckOptions,
  Command,
  DependencyType,
  ExcludedPackage,
  ManifestProfile,
  MetaPackage,
  Package,
  Profile,
  ProfilesInfo,
  ProfileVersion,
  VigilantOptions,
} from "./types";
