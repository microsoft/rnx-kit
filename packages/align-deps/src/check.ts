import type { KitConfig } from "@rnx-kit/config";
import { getKitCapabilities, getKitConfig } from "@rnx-kit/config";
import { error, info, warn } from "@rnx-kit/console";
import { isPackageManifest, readPackage } from "@rnx-kit/tools-node/package";
import chalk from "chalk";
import { diffLinesUnified } from "jest-diff";
import path from "path";
import { migrateConfig } from "./compatibility/config";
import { gatherRequirements, getRequirements } from "./dependencies";
import { findBadPackages } from "./findBadPackages";
import { modifyManifest } from "./helpers";
import { updatePackageManifest } from "./manifest";
import { filterPreset, mergePresets } from "./preset";
import { resolveCustomProfiles } from "./profiles";
import type {
  AlignDepsConfig,
  CheckConfig,
  CheckOptions,
  Command,
  ErrorCode,
  Options,
  Preset,
} from "./types";

export function containsValidPresets(config: KitConfig["alignDeps"]): boolean {
  const presets = config?.presets;
  return !presets || (Array.isArray(presets) && presets.length > 0);
}

export function containsValidRequirements(
  config: KitConfig["alignDeps"]
): boolean {
  const requirements = config?.requirements;
  if (requirements) {
    if (Array.isArray(requirements)) {
      return requirements.length > 0;
    } else if (typeof requirements === "object") {
      return (
        Array.isArray(requirements.production) &&
        requirements.production.length > 0
      );
    }
  }
  return false;
}

function ensurePreset(preset: Preset, requirements: string[]): void {
  if (Object.keys(preset).length === 0) {
    throw new Error(
      `No profiles could satisfy requirements: ${requirements.join(", ")}`
    );
  }
}

export function getCheckConfig(
  manifestPath: string,
  {
    loose,
    uncheckedReturnCode = 0,
    supportedVersions,
    targetVersion,
  }: CheckOptions
): number | CheckConfig {
  const manifest = readPackage(manifestPath);
  if (!isPackageManifest(manifest)) {
    error(
      `'${manifestPath}' does not contain a valid package manifest - please make sure it's not missing 'name' or 'version'`
    );
    return 1;
  }

  const badPackages = findBadPackages(manifest);
  if (badPackages) {
    warn(
      `Known bad packages are found in '${manifest.name}':\n` +
        badPackages
          .map((pkg) => `    ${pkg.name}@${pkg.version}: ${pkg.reason}`)
          .join("\n")
    );
  }

  const projectRoot = path.dirname(manifestPath);
  const kitConfig = getKitConfig({ cwd: projectRoot });
  if (!kitConfig) {
    return uncheckedReturnCode;
  }

  const {
    reactNativeVersion: targetReactNativeVersion,
    reactNativeDevVersion,
    kitType,
    capabilities: targetCapabilities,
    customProfiles,
  } = getKitCapabilities({
    // React Native versions declared in the package's config should always
    // override the ones specified with the `--vigilant` flag.
    ...(supportedVersions
      ? { reactNativeVersion: supportedVersions }
      : undefined),
    // We should not set dev version if the package is configured. It may have
    // been intentionally left out to reuse `reactNativeVersion`.
    ...(!kitConfig.reactNativeVersion && targetVersion
      ? { reactNativeDevVersion: targetVersion }
      : undefined),
    ...kitConfig,
  });

  const customProfilesPath = resolveCustomProfiles(projectRoot, customProfiles);

  const { reactNativeVersion, capabilities: requiredCapabilities } =
    getRequirements(
      targetReactNativeVersion,
      kitType,
      manifest,
      projectRoot,
      customProfilesPath,
      { loose }
    );
  requiredCapabilities.push(...targetCapabilities);

  return {
    kitType,
    reactNativeVersion,
    reactNativeDevVersion,
    capabilities: requiredCapabilities,
    customProfilesPath,
    manifest,
  };
}

export function getConfig(
  manifestPath: string
): AlignDepsConfig | CheckConfig | ErrorCode {
  const manifest = readPackage(manifestPath);
  if (!isPackageManifest(manifest)) {
    return "invalid-manifest";
  }

  const badPackages = findBadPackages(manifest);
  if (badPackages) {
    warn(
      `Known bad packages are found in '${manifest.name}':\n` +
        badPackages
          .map((pkg) => `\t${pkg.name}@${pkg.version}: ${pkg.reason}`)
          .join("\n")
    );
  }

  const projectRoot = path.dirname(manifestPath);
  const kitConfig = getKitConfig({ cwd: projectRoot });
  if (!kitConfig) {
    return "not-configured";
  }

  const { kitType = "library", alignDeps, ...config } = kitConfig;
  if (alignDeps) {
    const errors = [];
    if (!containsValidPresets(alignDeps)) {
      errors.push(`${manifestPath}: 'alignDeps.presets' cannot be empty`);
    }
    if (!containsValidRequirements(alignDeps)) {
      errors.push(`${manifestPath}: 'alignDeps.requirements' cannot be empty`);
    }
    if (errors.length > 0) {
      for (const e of errors) {
        error(e);
      }
      return "invalid-configuration";
    }
    return {
      kitType,
      alignDeps: {
        presets: ["microsoft/react-native"],
        requirements: [],
        capabilities: [],
        ...alignDeps,
      },
      ...config,
      manifest,
    };
  }

  const {
    capabilities,
    customProfiles,
    reactNativeDevVersion,
    reactNativeVersion,
  } = getKitCapabilities(config);

  return {
    kitType,
    reactNativeVersion,
    ...(config.reactNativeDevVersion ? { reactNativeDevVersion } : undefined),
    capabilities,
    customProfiles,
    manifest,
  } as CheckConfig;
}

function resolve(
  { kitType, alignDeps, manifest }: AlignDepsConfig,
  projectRoot: string,
  options: CheckOptions
) {
  const { capabilities, presets, requirements } = alignDeps;

  const prodRequirements = Array.isArray(requirements)
    ? requirements
    : requirements.production;
  const mergedPreset = mergePresets(presets, projectRoot);
  const initialProdPreset = filterPreset(prodRequirements, mergedPreset);
  ensurePreset(initialProdPreset, prodRequirements);

  const devPreset = (() => {
    if (kitType === "app") {
      // Preset for development is unused when the package is an app.
      return {};
    } else if (Array.isArray(requirements)) {
      return initialProdPreset;
    } else {
      const devRequirements = requirements.development;
      const devPreset = filterPreset(devRequirements, mergedPreset);
      ensurePreset(devPreset, devRequirements);
      return devPreset;
    }
  })();

  if (kitType === "app") {
    const { preset: prodMergedPreset, capabilities: mergedCapabilities } =
      gatherRequirements(
        projectRoot,
        manifest,
        initialProdPreset,
        capabilities,
        options
      );
    return {
      devPreset,
      prodPreset: prodMergedPreset,
      capabilities: mergedCapabilities,
    };
  }

  return { devPreset, prodPreset: initialProdPreset, capabilities };
}

export function checkPackageManifest(
  manifestPath: string,
  options: CheckOptions,
  inputConfig = getConfig(manifestPath)
): ErrorCode {
  if (typeof inputConfig === "string") {
    return inputConfig;
  }

  const config = migrateConfig(inputConfig);
  const { devPreset, prodPreset, capabilities } = resolve(
    config,
    path.dirname(manifestPath),
    options
  );
  if (capabilities.length === 0) {
    return "success";
  }

  const { kitType, manifest } = config;

  if (kitType === "app") {
    info(
      "Aligning dependencies according to the following profiles:",
      Object.keys(prodPreset).join(", ")
    );
  } else {
    info("Aligning dependencies according to the following profiles:");
    info("\t- Development:", Object.keys(devPreset).join(", "));
    info("\t- Production:", Object.keys(prodPreset).join(", "));
  }

  const updatedManifest = updatePackageManifest(
    manifest,
    capabilities,
    Object.values(prodPreset),
    Object.values(devPreset),
    kitType
  );

  // Don't fail when manifests only have whitespace differences.
  const updatedManifestJson = JSON.stringify(updatedManifest, undefined, 2);
  const normalizedManifestJson = JSON.stringify(manifest, undefined, 2);

  if (updatedManifestJson !== normalizedManifestJson) {
    if (options.write) {
      modifyManifest(manifestPath, updatedManifest);
    } else {
      const diff = diffLinesUnified(
        normalizedManifestJson.split("\n"),
        updatedManifestJson.split("\n"),
        {
          aAnnotation: "Current",
          aColor: chalk.red,
          bAnnotation: "Expected",
          bColor: chalk.green,
        }
      );
      console.log(diff);
      return "unsatisfied";
    }
  }

  return "success";
}

export function makeCheckCommand(options: Options): Command {
  return (manifest: string) => checkPackageManifest(manifest, options);
}
