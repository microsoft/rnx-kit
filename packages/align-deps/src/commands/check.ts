import { info } from "@rnx-kit/console";
import { pickValues } from "@rnx-kit/tools-language";
import type { PackageManifest } from "@rnx-kit/tools-node/package";
import { readPackage } from "@rnx-kit/tools-node/package";
import chalk from "chalk";
import { diffLinesUnified } from "jest-diff";
import * as path from "path";
import { migrateConfig } from "../compatibility/config";
import { loadConfig, loadPresetsOverrideFromPackage } from "../config";
import { isError } from "../errors";
import { modifyManifest } from "../helpers";
import { updatePackageManifest } from "../manifest";
import { resolve } from "../preset";
import type { Command, ErrorCode, Options } from "../types";
import { checkPackageManifestUnconfigured } from "./vigilant";

const visibleKeys = [
  "name",
  "version",
  "dependencies",
  "peerDependencies",
  "devDependencies",
];

function stringify(manifest: PackageManifest): string {
  return JSON.stringify(pickValues(manifest, visibleKeys), undefined, 2);
}

export function checkPackageManifest(
  manifestPath: string,
  options: Options,
  inputConfig = loadConfig(manifestPath)
): ErrorCode {
  if (isError(inputConfig)) {
    return inputConfig;
  }

  const config = migrateConfig(inputConfig, manifestPath, options);
  const { devPreset, prodPreset, capabilities } = resolve(
    config,
    path.dirname(manifestPath),
    options
  );
  const { kitType, manifest } = config;

  if (kitType === "app" && Object.keys(prodPreset).length !== 1) {
    return "invalid-app-requirements";
  } else if (capabilities.length === 0) {
    return "success";
  }

  if (options.verbose) {
    if (kitType === "app") {
      info(
        `${manifestPath}: Aligning your app's dependencies according to the following profiles:`,
        Object.keys(prodPreset).join(", ")
      );
    } else {
      info(
        `${manifestPath}: Aligning your library's dependencies according to the following profiles:\n` +
          `\t- Development: ${Object.keys(devPreset).join(", ")}\n` +
          `\t- Production: ${Object.keys(prodPreset).join(", ")}`
      );
    }
  }

  const updatedManifest = updatePackageManifest(
    manifestPath,
    manifest,
    capabilities,
    prodPreset,
    devPreset,
    kitType
  );

  // Don't fail when manifests only have whitespace differences.
  const updatedManifestJson = stringify(updatedManifest);
  const normalizedManifestJson = stringify(manifest);

  if (updatedManifestJson !== normalizedManifestJson) {
    if (options.write) {
      modifyManifest(manifestPath, updatedManifest);
    } else {
      const diff = diffLinesUnified(
        normalizedManifestJson.split("\n"),
        updatedManifestJson.split("\n"),
        {
          aAnnotation: "Current",
          aColor: chalk.red,
          bAnnotation: "Expected",
          bColor: chalk.green,
        }
      );
      console.log(diff);
      return "unsatisfied";
    }
  }

  return "success";
}

export function makeCheckCommand(options: Options): Command {
  const { presets, requirements } = options;
  if (!requirements) {
    return (manifest: string) => checkPackageManifest(manifest, options);
  }

  return (manifest: string) => {
    const inputConfig = loadConfig(manifest, "vigilant");
    const config = isError(inputConfig)
      ? inputConfig
      : migrateConfig(inputConfig, manifest, options);

    // If the package is configured, run the normal check first.
    if (!isError(config)) {
      const result = checkPackageManifest(manifest, options, config);
      return result !== "success"
        ? result
        : checkPackageManifestUnconfigured(manifest, options, config);
    }

    // Otherwise, run the unconfigured check only.
    if (config === "invalid-configuration" || config === "not-configured") {
      // In "vigilant" mode, we allow packages to declare which presets should
      // be used in config, overriding the `--presets` flag.
      const presetsOverride = loadPresetsOverrideFromPackage(manifest);
      return checkPackageManifestUnconfigured(manifest, options, {
        kitType: "library",
        alignDeps: {
          presets: presetsOverride || presets,
          requirements,
          capabilities: [],
        },
        manifest: readPackage(manifest),
      });
    }

    return config;
  };
}
