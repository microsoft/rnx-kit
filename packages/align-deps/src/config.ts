import type { KitConfig } from "@rnx-kit/config";
import { getKitCapabilities, getKitConfig } from "@rnx-kit/config";
import { error, warn } from "@rnx-kit/console";
import { isPackageManifest, readPackage } from "@rnx-kit/tools-node/package";
import * as path from "path";
import { findBadPackages } from "./findBadPackages";
import type { AlignDepsConfig, ErrorCode, LegacyCheckConfig } from "./types";

type ConfigResult = AlignDepsConfig | LegacyCheckConfig | ErrorCode;

export const defaultConfig: AlignDepsConfig["alignDeps"] = {
  presets: ["microsoft/react-native"],
  requirements: [],
  capabilities: [],
};

export function containsValidPresets(config: KitConfig["alignDeps"]): boolean {
  const presets = config?.presets;
  return !presets || (Array.isArray(presets) && presets.length > 0);
}

export function containsValidRequirements(
  config: KitConfig["alignDeps"]
): boolean {
  const requirements = config?.requirements;
  if (requirements) {
    if (Array.isArray(requirements)) {
      return requirements.length > 0;
    } else if (typeof requirements === "object") {
      return (
        Array.isArray(requirements.production) &&
        requirements.production.length > 0
      );
    }
  }
  return false;
}

/**
 * Loads configuration from the specified package manifest.
 * @param manifestPath The path to the package manifest to load configuration from
 * @returns The configuration; otherwise an error code
 */
export function loadConfig(manifestPath: string): ConfigResult {
  const manifest = readPackage(manifestPath);
  if (!isPackageManifest(manifest)) {
    return "invalid-manifest";
  }

  const badPackages = findBadPackages(manifest);
  if (badPackages) {
    warn(
      `Known bad packages are found in '${manifestPath}':\n` +
        badPackages
          .map((pkg) => `\t${pkg.name}@${pkg.version}: ${pkg.reason}`)
          .join("\n")
    );
  }

  const projectRoot = path.dirname(manifestPath);
  const kitConfig = getKitConfig({ cwd: projectRoot });
  if (!kitConfig) {
    return "not-configured";
  }

  const { kitType = "library", alignDeps, ...config } = kitConfig;
  if (alignDeps) {
    const errors = [];
    if (!containsValidPresets(alignDeps)) {
      errors.push(`${manifestPath}: 'alignDeps.presets' cannot be empty`);
    }
    if (!containsValidRequirements(alignDeps)) {
      errors.push(`${manifestPath}: 'alignDeps.requirements' cannot be empty`);
    }
    if (errors.length > 0) {
      for (const e of errors) {
        error(e);
      }
      return "invalid-configuration";
    }
    return {
      kitType,
      alignDeps: {
        ...defaultConfig,
        ...alignDeps,
      },
      ...config,
      manifest,
    };
  }

  try {
    const {
      capabilities,
      customProfiles,
      reactNativeDevVersion,
      reactNativeVersion,
    } = getKitCapabilities(config);

    return {
      kitType,
      reactNativeVersion,
      ...(config.reactNativeDevVersion ? { reactNativeDevVersion } : undefined),
      capabilities,
      customProfiles,
      manifest,
    };
  } catch (e) {
    return "invalid-configuration";
  }
}
