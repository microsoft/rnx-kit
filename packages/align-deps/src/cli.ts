#!/usr/bin/env node

import { error } from "@rnx-kit/console";
import { isNonEmptyArray } from "@rnx-kit/tools-language/array";
import { hasProperty } from "@rnx-kit/tools-language/properties";
import { findPackageDir } from "@rnx-kit/tools-node/package";
import {
  findWorkspacePackages,
  findWorkspaceRoot,
} from "@rnx-kit/tools-workspaces";
import isString from "lodash/isString";
import * as path from "path";
import { makeCheckCommand } from "./commands/check";
import { makeInitializeCommand } from "./commands/initialize";
import { makeSetVersionCommand } from "./commands/setVersion";
import { defaultConfig } from "./config";
import { printError, printInfo } from "./errors";
import type { Args, Command } from "./types";

export const cliOptions = {
  "exclude-packages": {
    description:
      "Comma-separated list of package names to exclude from inspection.",
    type: "string",
    requiresArg: true,
  },
  init: {
    description:
      "Writes an initial kit config to the specified 'package.json'. Note that this only works for React Native packages.",
    choices: ["app", "library"],
    conflicts: ["requirements"],
  },
  loose: {
    default: false,
    description:
      "Determines how strict the React Native version requirement should be. Useful for apps that depend on a newer React Native version than their dependencies declare support for.",
    type: "boolean",
  },
  "migrate-config": {
    default: false,
    description:
      "Determines whether align-deps should try to update the config in 'package.json'.",
    type: "boolean",
  },
  presets: {
    description:
      "Comma-separated list of presets. This can be names to built-in presets, or paths to external presets.",
    type: "string",
    requiresArg: true,
  },
  requirements: {
    description:
      "Comma-separated list of requirements to apply if a package is not configured for align-deps.",
    type: "string",
    requiresArg: true,
  },
  "set-version": {
    description:
      "Sets `react-native` requirements for any configured package. There is an interactive prompt if no value is provided. The value should be a comma-separated list of `react-native` versions to set, where the first number specifies the development version. Example: `0.70,0.69`",
    type: "string",
    conflicts: ["init", "requirements"],
  },
  verbose: {
    default: false,
    description: "Increase logging verbosity",
    type: "boolean",
  },
  write: {
    default: false,
    description: "Writes changes to the specified 'package.json'.",
    type: "boolean",
  },
};

async function getManifests(
  packages: (string | number)[] | undefined
): Promise<string[] | undefined> {
  const cwd = process.cwd();
  // When positional arguments are not provided, we will get `undefined` if
  // invoked directly, and an empty array if invoked via
  // `@react-native-community/cli`.
  if (isNonEmptyArray(packages)) {
    return packages.reduce<string[]>((result, pkg) => {
      const dir = findPackageDir(pkg.toString());
      if (dir) {
        const pkgJson = path.join(dir, "package.json");
        const relativePath = path.relative(cwd, pkgJson);
        result.push(relativePath);
      }
      return result;
    }, []);
  }

  const packageDir = findPackageDir();
  if (!packageDir) {
    return undefined;
  }

  // Make sure we don't return all packages when run inside a package that just
  // happens to be part of a workspace.
  const currentPackageJson = path.join(packageDir, "package.json");
  const manifestPath = path.relative(cwd, currentPackageJson);
  try {
    if ((await findWorkspaceRoot()) !== packageDir) {
      return [manifestPath];
    }
  } catch (_) {
    return [manifestPath];
  }

  try {
    const allPackages = (await findWorkspacePackages()).map((p) =>
      path.join(path.relative(cwd, p), "package.json")
    );
    allPackages.push(manifestPath);
    return allPackages;
  } catch (e) {
    if (hasProperty(e, "message")) {
      error(e.message);
      return undefined;
    }

    throw e;
  }
}

function reportConflicts(conflicts: [string, string][], args: Args): boolean {
  return conflicts.reduce<boolean>((result, [lhs, rhs]) => {
    if (lhs in args && rhs in args) {
      error(`--${lhs} and --${rhs} cannot both be specified at the same time.`);
      return true;
    }
    return result;
  }, false);
}

async function makeCommand(args: Args): Promise<Command | undefined> {
  const conflicts: [string, string][] = [
    ["init", "set-version"],
    ["init", args.write ? "write" : "no-write"],
    ["set-version", args.write ? "write" : "no-write"],
  ];
  if (reportConflicts(conflicts, args)) {
    return undefined;
  }

  const {
    "exclude-packages": excludePackages,
    init,
    loose,
    "migrate-config": migrateConfig,
    presets,
    requirements,
    "set-version": setVersion,
    verbose,
    write,
  } = args;

  const options = {
    presets: presets?.toString()?.split(",") ?? defaultConfig.presets,
    loose,
    migrateConfig,
    verbose,
    write,
    excludePackages: excludePackages?.toString()?.split(","),
    requirements: requirements?.toString()?.split(","),
  };

  if (typeof init !== "undefined") {
    return makeInitializeCommand(init, options);
  }

  // When `--set-version` is without a value, `setVersion` is an empty string if
  // invoked directly. When invoked via `@react-native-community/cli`,
  // `setVersion` is `true` instead.
  if (setVersion || isString(setVersion)) {
    return makeSetVersionCommand(setVersion, options);
  }

  return makeCheckCommand(options);
}

export async function cli({ packages, ...args }: Args): Promise<void> {
  const command = await makeCommand(args);
  if (!command) {
    process.exitCode = 1;
    return;
  }

  const manifests = await getManifests(packages);
  if (!manifests) {
    error("Could not find package root");
    process.exitCode = 1;
    return;
  }

  // We will optimistically run through all packages regardless of failures. In
  // most scenarios, this should be fine: Both init and check+write write to
  // disk only when everything is in order for the target package. Packages with
  // invalid or missing configurations are skipped.
  const errors = manifests.reduce((errors, manifest) => {
    try {
      const result = command(manifest);
      if (result !== "success") {
        printError(manifest, result);
        return errors + 1;
      }
    } catch (e) {
      if (hasProperty(e, "message")) {
        error(`${manifest}: ${e.message}`);
        return errors + 1;
      }

      throw e;
    }
    return errors;
  }, 0);

  process.exitCode = errors;

  if (errors > 0) {
    printInfo();
  }
}

if (require.main === module) {
  require("yargs").usage(
    "$0 [packages...]",
    "Manage dependencies within a repository and across many repositories",
    cliOptions,
    cli
  ).argv;
}
