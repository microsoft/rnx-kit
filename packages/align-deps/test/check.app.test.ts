import path from "path";
import { checkPackageManifest } from "../src/commands/check";
import { defaultConfig } from "../src/config";

jest.mock("fs");
jest.unmock("@rnx-kit/config");

const defaultOptions = {
  presets: defaultConfig.presets,
  loose: false,
  migrateConfig: false,
  verbose: false,
  write: true,
};

function fixturePath(name: string) {
  return path.join(process.cwd(), "test", "__fixtures__", name);
}

describe("checkPackageManifest({ kitType: 'app' })", () => {
  test("fails if multiple profiles are returned", () => {
    const manifestPath = path.join(
      fixturePath("misconfigured-app"),
      "package.json"
    );

    const result = checkPackageManifest(manifestPath, defaultOptions);

    expect(result).toBe("invalid-app-requirements");
  });
});

describe("checkPackageManifest({ kitType: 'app' }) (backwards compatibility)", () => {
  const fs = require("fs");

  afterAll(() => {
    jest.clearAllMocks();
  });

  test("adds required dependencies", () => {
    const manifestPath = path.join(fixturePath("awesome-repo"), "package.json");

    let destination = "";
    let updatedManifest = "";
    fs.__setMockFileWriter((dest, content) => {
      destination = dest;
      updatedManifest = content;
    });

    expect(checkPackageManifest(manifestPath, defaultOptions)).toBe("success");
    expect(destination).toBe(manifestPath);
    expect(updatedManifest).toMatchSnapshot();
  });
});
