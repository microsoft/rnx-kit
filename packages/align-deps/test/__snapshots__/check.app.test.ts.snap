// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`checkPackageManifest({ kitType: 'app' }) (backwards compatibility) adds required dependencies 1`] = `
"{
  \\"name\\": \\"awesome-repo\\",
  \\"version\\": \\"1.0.0\\",
  \\"dependencies\\": {
    \\"@react-native-async-storage/async-storage\\": \\"^1.17.10\\",
    \\"@react-native-community/netinfo\\": \\"^9.0.0\\",
    \\"@rnx-kit/react-native-lazy-index\\": \\"^2.1.7\\",
    \\"conan\\": \\"1.0.0\\",
    \\"hermes-engine\\": \\"~0.11.0\\",
    \\"quaid\\": \\"1.0.0\\",
    \\"react\\": \\"18.1.0\\",
    \\"react-native\\": \\"^0.70.0\\",
    \\"react-native-reanimated\\": \\"^2.10.0\\",
    \\"react-native-webview\\": \\"^11.23.0\\"
  },
  \\"rnx-kit\\": {
    \\"reactNativeVersion\\": \\"^0.70\\",
    \\"kitType\\": \\"app\\",
    \\"capabilities\\": [
      \\"core-android\\",
      \\"hermes\\",
      \\"lazy-index\\"
    ]
  }
}
"
`;
